File: framework/languages/dotnet/e2e/RuntimeMonitoring/Client/Scenarios/MonD1FailureRecoveryScenario.cs
// Verifies MON-D1 public validation and repeated failure recovery behavior.
using System.Diagnostics;
using System.Net.Sockets;
using RuntimeMonitoring.Client.Support;
using RuntimeMonitoring.Shared;
using Zlink.Framework.E2E.Configuration;
using Zlink.HttpClient;

namespace RuntimeMonitoring.Client.Scenarios;

internal static class MonD1FailureRecoveryScenario
{
    public static async Task RunValidationAsync(ClientOptions options)
    {
        using var observer = ZLinkHttpClient.Create(options.ServiceUrl)
            .Timeout(TimeSpan.FromSeconds(40))
            .Build();
        var validation =
            (await observer.Get("/runtime/validate")
                .Async<RuntimeValidationRes>()).Body;
        ZlinkStreamAssert.Ensure(
            validation.MissingSnapshotRejected
            && validation.MissingObserverRejected
            && validation.RegisteredObserverProducedStatus,
            "MON-D1 public snapshot or observer validation accepted an invalid call.");

        Console.WriteLine("scenario MON-D1A passed");
    }

    public static async Task RunCrashRecoveryAsync(ClientOptions options)
    {
        using var observer = ZLinkHttpClient.Create(options.ServiceUrl)
            .Timeout(TimeSpan.FromSeconds(40))
            .Build();

        var serviceBUri = new Uri(options.ServiceBUrl);
        var serviceBChannelUri = new Uri(options.ServiceBChannelEndpoint);
        Process? current = Process.GetProcessById(options.ServiceBProcessId);
        var lastSequence = 0UL;
        try
        {
            for (var cycle = 1; cycle <= 3; cycle++)
            {
                var readyBefore = await WaitForReadyPeerAsync(observer, "svc-b");
                var baseline =
                    (await observer.Get("/evidence").Async<string[]>()).Body.Length;

                current.Kill(entireProcessTree: true);
                await current.WaitForExitAsync()
                    .WaitAsync(TimeSpan.FromSeconds(10));
                current.Dispose();
                current = null;
                await WaitForPortStateAsync(
                    serviceBUri.Host,
                    serviceBUri.Port,
                    false,
                    $"MON-D1 cycle {cycle} expected service-b HTTP port to close.");
                var unavailable = await WaitUntilNotReadyAsync(
                    observer,
                    "svc-b");
                ZlinkStreamAssert.Ensure(
                    unavailable.Sequence > readyBefore.Sequence
                    && !unavailable.Peers.Any(peer =>
                        peer.Rid.StartsWith("svc-b-", StringComparison.Ordinal)
                        && peer.State == "Ready"),
                    $"MON-D1 cycle {cycle} retained the failed peer as ready.");

                current = StartServiceB(options, cycle);
                await WaitForPortStateAsync(
                    serviceBUri.Host,
                    serviceBUri.Port,
                    true,
                    $"MON-D1 cycle {cycle} expected service-b to restart.");
                await WaitForPortStateAsync(
                    serviceBChannelUri.Host,
                    serviceBChannelUri.Port,
                    true,
                    $"MON-D1 cycle {cycle} expected the channel endpoint to restart.");
                var recovered = await WaitForReadyPeerAsync(observer, "svc-b");
                ZlinkStreamAssert.Ensure(
                    recovered.Sequence > unavailable.Sequence
                    && recovered.Peers.Count(peer =>
                        peer.Rid.StartsWith("svc-b-", StringComparison.Ordinal)
                        && peer.State == "Ready") == 1
                    && recovered.Channels.Any(channel =>
                        channel.ChannelName == RuntimeMonitoringNames.Channel
                        && channel.IsReady),
                    $"MON-D1 cycle {cycle} snapshot did not resync to the replacement peer.");

                using var activeServiceB =
                    ZLinkHttpClient.Create(options.ServiceBUrl)
                        .Timeout(TimeSpan.FromSeconds(35))
                        .Build();
                await activeServiceB.Post("/admin/weight/exclude")
                    .Async<object>();
                await Task.Delay(150);
                await activeServiceB.Post("/admin/weight/include")
                    .Async<object>();

                var cycleEvidence = (await observer.Post("/evidence/wait")
                    .Body(new EvidenceWaitReq(
                        [
                            $"source={RuntimeMonitoringNames.Channel}",
                            "identifier=zlink.runtime.mesh_node.peer_changed",
                            "identifier=zlink.runtime.mesh_node.channel_changed",
                            "routing=svc-b"
                        ],
                        [],
                        TimeoutMilliseconds: 3000,
                        AfterIndex: baseline))
                    .Async<string[]>()).Body;
                var eventLines = cycleEvidence.Where(line =>
                        line.Contains(
                            $"source={RuntimeMonitoringNames.Channel}",
                            StringComparison.Ordinal)
                        && line.Contains(
                            "identifier=zlink.runtime.",
                            StringComparison.Ordinal))
                    .ToArray();
                var sequences = eventLines
                    .Select(ParseSequence)
                    .Where(static sequence => sequence > 0)
                    .Distinct()
                    .ToArray();
                ZlinkStreamAssert.Ensure(
                    sequences.Length >= 2
                    && sequences.All(sequence => sequence > lastSequence)
                    && sequences.Zip(
                            sequences.Skip(1),
                            static (left, right) => right > left)
                        .All(static increasing => increasing),
                    $"MON-D1 cycle {cycle} event sequence was not strictly increasing.");
                lastSequence = sequences[^1];
                ZlinkStreamAssert.Ensure(
                    eventLines.All(line =>
                        !line.Contains("mon-d1-payload", StringComparison.Ordinal)
                        && !line.Contains("profile-request|", StringComparison.Ordinal)
                        && !line.Contains("|value=", StringComparison.Ordinal)),
                    $"MON-D1 cycle {cycle} copied application metadata into a runtime event.");
            }

            await observer.Post("/admin/weight/exclude").Async<object>();
            using var finalServiceB =
                ZLinkHttpClient.Create(options.ServiceBUrl)
                    .Timeout(TimeSpan.FromSeconds(35))
                    .Build();
            var reply = (await finalServiceB.Post("/profile/request")
                .Body(new ProfileReq("restart", "mon-d1-payload"))
                .Async<ProfileRes>()).Body;
            ZlinkStreamAssert.Ensure(
                reply.ProviderRid == "svc-b"
                && reply.Marker == "mon-d1-payload"
                && reply.Value == "profile:restart",
                "MON-D1 messaging did not recover after three failure cycles.");
        }
        finally
        {
            await PostBestEffortAsync(observer, "/admin/weight/include");
            using var activeServiceB = ZLinkHttpClient.Create(options.ServiceBUrl)
                .Timeout(TimeSpan.FromSeconds(20))
                .Build();
            await PostBestEffortAsync(activeServiceB, "/shutdown");
            if (current is not null)
            {
                try
                {
                    await current.WaitForExitAsync()
                        .WaitAsync(TimeSpan.FromSeconds(5));
                }
                catch (TimeoutException)
                {
                    if (!current.HasExited)
                        current.Kill(true);
                    await current.WaitForExitAsync();
                }
                current.Dispose();
            }
        }

        Console.WriteLine("scenario MON-D1B passed");
    }

    private static Process StartServiceB(ClientOptions options, int cycle)
    {
        var stdout = Path.Combine(
            options.LogDir,
            $"svc-b-restart-{cycle}.stdout.log");
        var stderr = Path.Combine(
            options.LogDir,
            $"svc-b-restart-{cycle}.stderr.log");
        var startInfo = new ProcessStartInfo("dotnet")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false
        };
        startInfo.ArgumentList.Add("run");
        startInfo.ArgumentList.Add("--no-build");
        startInfo.ArgumentList.Add("--project");
        startInfo.ArgumentList.Add(options.ServiceProject);
        startInfo.ArgumentList.Add("--");
        startInfo.ArgumentList.Add("--config");
        startInfo.ArgumentList.Add(E2eConfiguration.Write(
            options.ConfigDir,
            $"svc-b-restart-{cycle}",
            new RestartServiceOptions(
                "service",
                options.ServiceBUrl,
                options.LogDir,
                "svc-b",
                Path.Combine(
                    options.LogDir,
                    $"svc-b-restart-{cycle}.evidence.log"),
                options.RedisEndpoint,
                options.RedisKeyPrefix,
                options.ServiceBChannelEndpoint,
                options.ServiceBSpotRouterEndpoint,
                options.ServiceBSpotPubEndpoint)));

        var process = Process.Start(startInfo)
                      ?? throw new InvalidOperationException(
                          "Failed to restart service-b.");
        _ = Task.Run(async () =>
            await File.WriteAllTextAsync(
                stdout,
                await process.StandardOutput.ReadToEndAsync()));
        _ = Task.Run(async () =>
            await File.WriteAllTextAsync(
                stderr,
                await process.StandardError.ReadToEndAsync()));
        return process;
    }

    private static async Task<MeshRuntimeSnapshotRes> SnapshotAsync(
        ZLinkHttpClient service)
        => (await service.Get(
                $"/runtime/snapshot/{RuntimeMonitoringNames.Channel}")
            .Async<MeshRuntimeSnapshotRes>()).Body;

    private static async Task<MeshRuntimeSnapshotRes> WaitForReadyPeerAsync(
        ZLinkHttpClient service,
        string rid)
    {
        var elapsed = Stopwatch.StartNew();
        while (true)
        {
            var snapshot = await SnapshotAsync(service);
            if (snapshot.Peers.Any(peer =>
                    peer.Rid.StartsWith($"{rid}-", StringComparison.Ordinal)
                    && peer.State == "Ready"))
                return snapshot;
            if (elapsed.Elapsed >= TimeSpan.FromSeconds(15))
                throw new InvalidOperationException(
                    $"MON-D1 peer '{rid}' did not become ready.");
            await Task.Delay(50);
        }
    }

    private static async Task<MeshRuntimeSnapshotRes> WaitUntilNotReadyAsync(
        ZLinkHttpClient service,
        string rid)
    {
        var elapsed = Stopwatch.StartNew();
        while (true)
        {
            var snapshot = await SnapshotAsync(service);
            if (!snapshot.Peers.Any(peer =>
                    peer.Rid.StartsWith($"{rid}-", StringComparison.Ordinal)
                    && peer.State == "Ready"))
                return snapshot;
            if (elapsed.Elapsed >= TimeSpan.FromSeconds(15))
                throw new InvalidOperationException(
                    $"MON-D1 peer '{rid}' remained ready.");
            await Task.Delay(50);
        }
    }

    private static ulong ParseSequence(string line)
    {
        const string prefix = "|sequence=";
        var start = line.IndexOf(prefix, StringComparison.Ordinal);
        if (start < 0)
            return 0;
        start += prefix.Length;
        var end = line.IndexOf('|', start);
        var text = end < 0 ? line[start..] : line[start..end];
        return ulong.TryParse(text, out var value) ? value : 0;
    }

    private static async Task PostBestEffortAsync(
        ZLinkHttpClient http,
        string path)
    {
        try
        {
            await http.Post(path).Async<object>();
        }
        catch (HttpRequestException)
        {
        }
        catch (TaskCanceledException)
        {
        }
        catch (Zlink.Framework.Contracts.Errors.ZLinkFrameworkException)
        {
        }
    }

    private static async Task WaitForPortStateAsync(
        string host,
        int port,
        bool shouldBeOpen,
        string failureMessage)
    {
        for (var attempt = 0; attempt < 30; attempt++)
        {
            if (await CanConnectAsync(host, port) == shouldBeOpen)
                return;
            await Task.Delay(100);
        }
        throw new InvalidOperationException(failureMessage);
    }

    private static async Task<bool> CanConnectAsync(string host, int port)
    {
        try
        {
            using var client = new TcpClient();
            await client.ConnectAsync(host, port)
                .WaitAsync(TimeSpan.FromMilliseconds(200));
            return true;
        }
        catch (SocketException)
        {
            return false;
        }
        catch (TimeoutException)
        {
            return false;
        }
    }
}

internal sealed record RestartServiceOptions(
    string Role,
    string HttpUrl,
    string LogDir,
    string Rid,
    string EvidenceFile,
    string RedisEndpoint,
    string RedisKeyPrefix,
    string ChannelEndpoint,
    string SpotRouterEndpoint,
    string SpotPubEndpoint);

File: framework/languages/node/e2e/RuntimeMonitoring/Client/Scenarios/mon-d1-failure-recovery-scenario.ts
// MON-D1: public validation과 장애 반복 중 연속성 시나리오를 검증한다.
import type { EvidenceWaitReq, ProfileReq, ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { type ManagedProcess, startServiceB, waitForPortState } from '../Support/managed-service';
import {
  routeStatusesFromEvidence,
  waitForRouteStatus,
  delay
} from '../Support/public-status';
import { ensure } from '../Support/scenario-assert';

export async function runMonD1(options: ClientOptions): Promise<ManagedProcess> {
  const result = await restartAndProbe(options, 'mon-d1-request');
  console.log('scenario MON-D1 passed');
  return result.process;
}

export async function runMonD1B(options: ClientOptions): Promise<ManagedProcess> {
  let current: ManagedProcess | undefined;
  try {
    for (let cycle = 1; cycle <= 3; cycle += 1) {
      const result = await restartAndProbe(options, `mon-d1b-cycle-${cycle}`);
      current = result.process;
    }
    console.log('scenario repeated restart passed');
    return current!;
  } catch (error) {
    await current?.stop();
    throw error;
  }
}

async function restartAndProbe(
  options: ClientOptions,
  marker: string
): Promise<{ readonly process: ManagedProcess; readonly routeSequence: string }> {
  const before = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.isReady && status.peers.some((peer) => peer.nodeRid === 'svc-b'),
    'MON-D1 expected svc-b to be ready before restart.'
  );
  await postJson<object>(options.serviceBUrl, '/shutdown', {});
  await waitForPortState(options.serviceBUrl, false, 'MON-D1 expected service-b to stop.');
  const removed = await waitForRouteStatus(
    options.serviceUrl,
    (status) => !status.peers.some((peer) => peer.nodeRid === 'svc-b'),
    'MON-D1 observer did not expose the removed peer.'
  );

  const restarted = startServiceB(options, `svc-b-${marker}`);
  try {
    await waitForPortState(options.serviceBUrl, true, 'MON-D1 expected service-b to restart.');
    const restored = await waitForRouteStatus(
      options.serviceUrl,
      (status) => status.isReady && status.peers.some((peer) => peer.nodeRid === 'svc-b')
        && status.readyPeerCount > removed.readyPeerCount,
      'MON-D1 observer did not expose the restarted peer.'
    );
    ensure(BigInt(restored.sequence) > BigInt(before.sequence), 'MON-D1 RouteMesh sequence did not increase after restart.');

    const reply = await postJson<ProfileRes>(options.triggerUrl, '/profile/request/service-b', {
      value: 'restart',
      marker
    } satisfies ProfileReq);
    ensure(reply.providerRid === 'svc-b' && reply.marker === marker && reply.value === 'profile:restart', 'MON-D1 restarted service did not handle request.');
    const evidence = await postJson<string[]>(options.serviceBUrl, '/evidence/wait', {
      containsAll: [`profile-request|rid=svc-b|marker=${marker}|value=restart`],
      containsAnyGroups: [],
      timeoutMilliseconds: 15000
    } satisfies EvidenceWaitReq);
    ensure(evidence.some((line) => line.includes(`profile-request|rid=svc-b|marker=${marker}|value=restart`)), 'MON-D1 restarted service evidence missing.');

    const observed = routeStatusesFromEvidence(await getJson<readonly string[]>(options.serviceUrl, '/evidence'));
    ensure(
      observed.some((status) => status.peers.every((peer) => peer.nodeRid !== 'svc-b'))
        && observed.some((status) => status.peers.some((peer) => peer.nodeRid === 'svc-b')),
      'MON-D1 observer did not retain a current status after peer restart.'
    );
    return { process: restarted, routeSequence: restored.sequence };
  } catch (error) {
    await restarted.stop();
    throw error;
  }
}

export async function waitForRouteObservation(options: ClientOptions, sequence: string): Promise<void> {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const statuses = routeStatusesFromEvidence(await getJson<readonly string[]>(options.serviceUrl, '/evidence'));
    if (statuses.some((status) => BigInt(status.sequence) >= BigInt(sequence))) return;
    await delay();
  }
  throw new Error('MON-D1 public RouteMesh observer did not catch up to the latest status.');
}
